src/main/java/uk/ac/ebi/ampt2d/commons/accession/generators/DecoratedAccessionGenerator.java

```diff
@@ -0,0 +1,95 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package uk.ac.ebi.ampt2d.commons.accession.generators;
+
+import uk.ac.ebi.ampt2d.commons.accession.core.SaveResponse;
+import uk.ac.ebi.ampt2d.commons.accession.generators.exceptions.AccessionCouldNotBeGeneratedException;
+import uk.ac.ebi.ampt2d.commons.accession.generators.monotonic.MonotonicAccessionGenerator;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class DecoratedAccessionGenerator<MODEL, ACCESSION> implements AccessionGenerator<MODEL, String> {
+
+    private AccessionGenerator<MODEL, ACCESSION> generator;
+
+    private Function<ACCESSION, String> decorateAccession;
+
+    private Function<String, ACCESSION> undecorateAccession;
+
+    public DecoratedAccessionGenerator(AccessionGenerator<MODEL, ACCESSION> generator,
+                                       Function<ACCESSION, String> decorateAccession,
+                                       Function<String, ACCESSION> undecorateAccession) {
+        this.generator = generator;
+        this.decorateAccession = decorateAccession;
+        this.undecorateAccession = undecorateAccession;
+    }
+
+    @Override
+    public <HASH> List<ModelHashAccession<MODEL, HASH, String>> generateAccessions(Map<HASH, MODEL> messages) throws AccessionCouldNotBeGeneratedException {
+        List<ModelHashAccession<MODEL, HASH, ACCESSION>> accessions = generator.generateAccessions(messages);
+        return accessions.stream()
+                .map(mha -> ModelHashAccession.of(mha.model(), mha.hash(), decorateAccession.apply(mha.accession())))
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public void postSave(SaveResponse<String, MODEL> response) {
+        Map<ACCESSION, MODEL> savedAccession = response.getSavedAccessions().entrySet().stream()
+                .collect(Collectors.toMap(o -> undecorateAccession.apply(o.getKey()), o -> o.getValue()));
+        Map<ACCESSION, MODEL> unsavedAccession = response.getUnsavedAccessions().entrySet().stream()
+                .collect(Collectors.toMap(o -> undecorateAccession.apply(o.getKey()), o -> o.getValue()));
+        generator.postSave(new SaveResponse<>(savedAccession, unsavedAccession));
+    }
+
+    public static <MODEL> DecoratedAccessionGenerator<MODEL, Long> prefixSuffixMonotonicAccessionGenerator(
+            MonotonicAccessionGenerator<MODEL> generator,
+            String prefix, String suffix) {
+        return prefixSuffixAccessionGenerator(generator, prefix, suffix, Long::parseLong);
+    }
+
+    public static <MODEL, ACCESSION> DecoratedAccessionGenerator<MODEL, ACCESSION> prefixSuffixAccessionGenerator(
```

Comment:
I mentioned this in a previous review but it got hidden after changing the packages: we don't really need to support suffixes so I could be keen on simplifying this a bit. But it is not a huge class anyway, so what do others think?

Reply:
I like the idea of having just what is needed. I prefer to invest the time later to add this if needed, that carrying this "dead code" forever if not needed.

Reply:
To me both cases are part of the same use case that is why I opted for a single generic implementation even if the eva is not gonna use suffixes.